x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts

```diff
@@ -20,25 +23,119 @@ export type Credential = Record<string, any>;
  */
 export interface CredentialStore {
   get(reindexOp: ReindexSavedObject): Credential | undefined;
-  set(reindexOp: ReindexSavedObject, credential: Credential): void;
+  set(params: {
+    reindexOp: ReindexSavedObject;
+    request: KibanaRequest;
+    security?: SecurityPluginStart;
+  }): Promise<void>;
+  update(params: {
+    reindexOp: ReindexSavedObject;
+    security?: SecurityPluginStart;
+    credential: Credential;
+  }): Promise<void>;
   clear(): void;
 }
 
-export const credentialStoreFactory = (): CredentialStore => {
+export const credentialStoreFactory = (logger: Logger): CredentialStore => {
   const credMap = new Map<string, Credential>();
+  const apiKeysMap = new Map<string, string>();
+  const log = logger.get('credential_store');
 
   // Generates a stable hash for the reindex operation's current state.
   const getHash = (reindexOp: ReindexSavedObject) =>
     createHash('sha256')
       .update(stringify({ id: reindexOp.id, ...reindexOp.attributes }))
       .digest('base64');
 
+  const getApiKey = async ({
+    request,
+    security,
+    reindexOpId,
+  }: {
+    request: KibanaRequest;
+    security?: SecurityPluginStart;
+    reindexOpId: string;
+  }): Promise<string | undefined> => {
+    const apiKeyResult = await security?.authc.apiKeys.grantAsInternalUser(request, {
+      name: `ua_reindex_${reindexOpId}`,
+      role_descriptors: {},
+    });
```

Comment:
Optional nit: given the chance that an API key could get orphaned, maybe it's worth adding a bit of metadata? Something like this:

```diff
diff --git a/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts b/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts
index 7f3f4f2b307..41eb7a5afa7 100644
--- a/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts
+++ b/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts
@@ -59,6 +59,10 @@ export const credentialStoreFactory = (logger: Logger): CredentialStore => {
     const apiKeyResult = await security?.authc.apiKeys.grantAsInternalUser(request, {
       name: `ua_reindex_${reindexOpId}`,
       role_descriptors: {},
+      metadata: {
+        description:
+          'Created by the Upgrade Assistant for a reindex operation; this can be safely deleted after Kibana is upgraded.',
+      },
     });
 
     if (apiKeyResult) {
```

Reply:
Great point! Done.